pkg/reconciler/buildrun/controller.go

```diff
@@ -71,7 +71,7 @@ func add(ctx context.Context, mgr manager.Manager, r reconcile.Reconciler, maxCo
 			// - when a BuildRun already have a referenced TaskRun, but the latest version is not canceled
 			// - when a BuildRun have a completionTime set
 			switch {
-			case o.GetLabels()[buildv1alpha1.LabelBuild] == "":
+			case o.Spec.BuildRef != nil && o.GetLabels()[buildv1alpha1.LabelBuild] == "":
```

Comment:
To better represent what is happening for "the build.shipwright.io/name label is set", would that also work?

```suggestion
			case o.GetLabels()[buildv1alpha1.LabelBuild] == "" && n.GetLabels()[buildv1alpha1.LabelBuild] != "":
```

Reply:
Does `n.GetLabels()[buildv1alpha1.LabelBuild] != "` do any meaningful restriction?

cases: 
- o is empty and n is empty -> would reconcile in your version and change the logic for existing brs with buildref
- o is empty and n is not empty -> don't reconcile, same as previous
- o is not empty and n is empty -> not possible

Seems like the labels on n do not matter or diverges the behaviour.

Reply:
We just paired on this predicate and think it is unnecessarily complex. The only use cases we have at the moment where we must reconcile on a BuildRun update, are these two:

1. When .spec.state is set to BuildRunCanceled
2. When .spec.retention is changed

The second use case is handled by the separate BuildRunTTLCleanup controller. That makes use case (1) the only one where we must reconcile based on an update to a BuildRun. This would mean this:

```go
		UpdateFunc: func(e event.UpdateEvent) bool {
			// Ignore updates to CR status in which case metadata.Generation does not change
			o := e.ObjectOld.(*buildv1alpha1.BuildRun)
			n := e.ObjectNew.(*buildv1alpha1.BuildRun)

			// Only reconcile a BuildRun update when
			// - it is set to canceled
			switch {
			case !o.IsCanceled() && n.IsCanceled():
				return true
			}

			return false
		},
```

All other "workflow-driven" updates all come from the TaskRun. The initial reconcile to trigger the TaskRun creation comes from the BuildRun create event. @qu1queee are we missing something ?

Reply:
I don't think so. To be on the safe side you could leave ` o.GetGeneration() != n.GetGeneration()` instead of False, which is mainly there to avoid reconciling after updates on `.status`.  But in theory, nothing else beyond `.status` updates should take place and should be able to reconcile again.